plugins/SlicerT/WaveForm.cpp

```diff
@@ -0,0 +1,333 @@
+
+	// draw text if no sample loaded
+	if (m_currentSample.frames() < 2048)
+	{
+		brush.setPen(m_playHighlighColor);
+		brush.setFont(QFont(brush.font().family(), 9.0f, -1, false));
+		brush.drawText(
+			m_editorWidth / 2 - 100, m_editorHeight / 2 - 100, 200, 200, Qt::AlignCenter, tr("Drag sample to load"));
+		return;
+	}
+
+	// editor boundaries
+	float startFrame = m_seekerStart * m_currentSample.frames();
+	float endFrame = m_seekerEnd * m_currentSample.frames();
+	float numFramesToDraw = endFrame - startFrame;
+
+	// 0 centered line
+	brush.setPen(m_playHighlighColor);
+	brush.drawLine(0, m_editorHeight / 2, m_editorWidth, m_editorHeight / 2);
+
+	// draw waveform
+	brush.setPen(m_waveformColor);
+	float zoomOffset = ((float)m_editorHeight - m_zoomLevel * m_editorHeight) / 2;
+	m_currentSample.visualize(
+		brush, QRect(0, zoomOffset, m_editorWidth, m_zoomLevel * m_editorHeight), startFrame, endFrame);
+
+	// draw slicepoints
+	brush.setPen(QPen(m_sliceColor, 2));
+	for (int i = 0; i < m_slicePoints.size(); i++)
+	{
+		float xPos = (float)(m_slicePoints[i] - startFrame) / numFramesToDraw * m_editorWidth;
+
+		if (i == m_sliceSelected) { brush.setPen(QPen(m_selectedSliceColor, 2)); }
+		else { brush.setPen(QPen(m_sliceColor, 2)); }
+
+		brush.drawLine(xPos, 0, xPos, m_editorHeight);
+		brush.drawPixmap(xPos - (float)m_sliceArrow.width() / 2, 0, m_sliceArrow);
+	}
+}
+
+void WaveForm::updateUI()
+{
+	drawSeekerWaveform();
+	drawSeeker();
+	drawEditor();
+	update();
+}
+
+void WaveForm::isPlaying(float current, float start, float end)
+{
+	m_noteCurrent = current;
+	m_noteStart = start;
+	m_noteEnd = end;
+	drawSeeker(); // only update seeker, else horrible performance because of waveform redraw
+	update();
+}
+
+// events
+void WaveForm::mousePressEvent(QMouseEvent* me)
+{
+	float normalizedClickSeeker = (float)(me->x() - m_seekerHorMargin) / m_seekerWidth;
+	float normalizedClickEditor = (float)(me->x()) / m_editorWidth;
+	// reset seeker on middle click
+	if (me->button() == Qt::MouseButton::MiddleButton)
+	{
+		m_seekerStart = 0;
+		m_seekerEnd = 1;
+		m_zoomLevel = 1;
+		return;
+	}
+
+	if (me->y() < m_seekerHeight) // seeker click
+	{
+		if (abs(normalizedClickSeeker - m_seekerStart) < m_distanceForClick) // dragging start
+		{
+			m_currentlyDragging = m_draggingTypes::m_seekerStart;
+		}
+		else if (abs(normalizedClickSeeker - m_seekerEnd) < m_distanceForClick) // dragging end
+		{
+			m_currentlyDragging = m_draggingTypes::m_seekerEnd;
+		}
+		else if (normalizedClickSeeker > m_seekerStart && normalizedClickSeeker < m_seekerEnd) // dragging middle
+		{
+			m_currentlyDragging = m_draggingTypes::m_seekerMiddle;
+			m_seekerMiddle = normalizedClickSeeker;
+		}
+	}
+	else // editor click
+	{
+		m_sliceSelected = -1;
+		float startFrame = m_seekerStart * m_currentSample.frames();
+		float endFrame = m_seekerEnd * m_currentSample.frames();
+		// select slice
+		for (int i = 0; i < m_slicePoints.size(); i++)
```

Comment:
The index is needed to set the selected slice, so I think it's fine like this. In general, is there any advantage to use range-based for loops? If not, I would like to keep the normal loops, just because I already use them everywhere in the plugin.

Reply:
> The index is needed to set the selected slice, so I think it's fine like this. In general, is there any advantage to use range-based for loops? If not, I would like to keep the normal loops, just because I already use them everywhere in the plugin.

You could still use a range based for loop to do that, just take the element by reference. However, I do not want to make this so overwhelming. Generally, in C++, when we iterate over collections, using a range based for loop [expresses intent](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#p3-express-intent) better. However, if you feel like it's too much and doesn't have a good ROI in terms of the time it would take for you to change them, I would leave it (or I could make the suggestions for you if you want).

Reply:
I don't care about the effort if the code is better, but I don't think that these loops fit anywhere in the current code. If I where to select the slice by reference, I would have to store it as a class member, which you said wasn't good for copying and stuff. Plus I would have to change a lot of code elsewhere where I use the indexes for comparisons, so overall I don't think its worth the effort for one for loop. If there are any loops where it works better, I would be happy to change it, or you can if you want.

Reply:
We can leave it then, not a problem. :+1: